src/wp-includes/query.php

```diff
@@ -1117,7 +1118,20 @@ function _find_post_by_old_slug( $post_type ) {
 		$query .= $wpdb->prepare( ' AND DAYOFMONTH(post_date) = %d', get_query_var( 'day' ) );
 	}
 
-	$id = (int) $wpdb->get_var( $query );
+	$key          = md5( $query );
+	$last_changed = wp_cache_get( 'last_changed', 'posts' );
+	if ( ! $last_changed ) {
+		$last_changed = microtime();
+		wp_cache_set( 'last_changed', $last_changed, 'posts' );
+	}
+	$cache_key = "_find_post_by_old_slug:$key:$last_changed";
+	$cache     = wp_cache_get( $cache_key, 'posts' );
+	if ( $cache !== false ) {
+		$id = $cache;
+	} else {
+		$id = (int) $wpdb->get_var( $query );
+		wp_cache_set( $cache_key, $id, 'posts' );
+	}
```

Comment:
```suggestion
	if ( false !== $cache ) {
		return $cache;
	}

	$id = (int) $wpdb->get_var( $query );
	wp_cache_set( $cache_key, $id, 'posts' );
```

I find fewer indents make the code easier to read but take this as a nitpick.